src/Orleans/Storage/GenericRelationalStorage.cs

```diff
@@ -0,0 +1,210 @@
+            {
+                throw new ArgumentException("Connection string must contain characters", "connectionString");
+            }
+
+            return new GenericRelationalStorage(invariantName, connectionString);
+        }
+
+
+        /// <summary>
+        /// Executes a given statement. Especially intended to use with <em>SELECT</em> statement.
+        /// </summary>
+        /// <typeparam name="TResult">The result type.</typeparam>
+        /// <param name="query">Executes a given statement. Especially intended to use with <em>SELECT</em> statement.</param>        
+        /// <param name="parameterProvider">Adds parameters to the query. Parameter names must match those defined in the query.</param>
+        /// <param name="selector">This function transforms the raw <see cref="IDataRecord"/> results to type <see paramref="TResult"/> the <see cref="int"/> parameter being the resultset number.</param>
+        /// <returns>A list of objects as a result of the <see paramref="query"/>.</returns>
+        public async Task<IEnumerable<TResult>> ReadAsync<TResult>(string query, Action<IDbCommand> parameterProvider, Func<IDataRecord, int, TResult> selector)
+        {
+            //If the query is something else that is not acceptable (e.g. an empty string), there will an appropriate database exception.
+            if(query == null)
+            {
+                throw new ArgumentNullException("query");
+            }
+
+            if(selector == null)
+            {
+                throw new ArgumentNullException("selector");
+            }
+
+            return (await ExecuteAsync(query, parameterProvider, ExecuteReaderAsync, selector, factory, connectionString).ConfigureAwait(continueOnCapturedContext: false)).Item1;
+        }
+
+
+        /// <summary>
+        /// Executes a given statement. Especially intended to use with <em>INSERT</em>, <em>UPDATE</em>, <em>DELETE</em> or <em>DDL</em> queries.
+        /// </summary>
+        /// <param name="query">The query to execute.</param>
+        /// <param name="parameterProvider">Adds parameters to the query. Parameter names must match those defined in the query.</param>
+        /// <returns>Affected rows count.</returns>
+        public async Task<int> ExecuteAsync(string query, Action<IDbCommand> parameterProvider)
+        {
+            //If the query is something else that is not acceptable (e.g. an empty string), there will an appropriate database exception.
+            if(query == null)
+            {
+                throw new ArgumentNullException("query");
+            }
+
+            return (await ExecuteAsync(query, parameterProvider, ExecuteReaderAsync, (unit, id) => unit, factory, connectionString).ConfigureAwait(continueOnCapturedContext: false)).Item2;
+        }
+
+
+        protected GenericRelationalStorage(string invariantName, string connectionString)
+        {
+            factory = DbProviderFactories.GetFactory(invariantName);
```

Comment:
Here goes the invariant name. ADO.NET ``DbProviderFactories`` loads the appropriate dlls with the given provider invariant name. To be able to do so, it needs to know of them somehow first, either through GAC or by other configuration elements (see eariler comments).